src/rust/engine/src/console_ui.rs

```diff
@@ -0,0 +1,192 @@
+// Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+// Licensed under the Apache License, Version 2.0 (see LICENSE).
+use crate::scheduler::RootResult;
+use indexmap::IndexMap;
+use indicatif::{MultiProgress, ProgressBar, ProgressStyle};
+use logging::logger::LOGGER;
+use parking_lot::Mutex;
+use std::sync::{mpsc, Arc, Barrier};
+use std::time::Duration;
+use workunit_store::WorkUnitStore;
+
+enum ConsoleMessage {
+  FinalResult,
+  //Stdout(String),
+  Stderr(String),
+  DisableUI,
+}
+
+#[derive(Clone)]
+pub struct ConsoleUI {
+  workunit_store: WorkUnitStore,
+  display_sender: Arc<Mutex<Option<mpsc::Sender<ConsoleMessage>>>>,
+  final_result: Arc<Mutex<Option<Vec<RootResult>>>>,
+  disable_ui_barrier: Arc<Barrier>,
+}
+
+impl ConsoleUI {
+  pub fn new(workunit_store: WorkUnitStore) -> ConsoleUI {
+    ConsoleUI {
+      workunit_store,
+      display_sender: Arc::new(Mutex::new(None)),
+      final_result: Arc::new(Mutex::new(None)),
+      disable_ui_barrier: Arc::new(Barrier::new(2)),
+    }
+  }
+
+  pub fn write_stdout(&self, msg: &str) {
+    self.with_console_ui_disabled(|| {
+      print!("{}", msg);
+    })
+  }
+
+  pub fn write_stderr(&self, msg: &str) {
+    if let Some(ref sender) = *self.display_sender.lock() {
+      sender
+        .send(ConsoleMessage::Stderr(msg.to_string()))
+        .unwrap();
+    }
+  }
+
+  pub fn with_console_ui_disabled<F: FnOnce() -> T, T>(&self, f: F) -> T {
+    if let Some(ref sender) = *self.display_sender.lock() {
+      sender.send(ConsoleMessage::DisableUI).unwrap();
+      self.disable_ui_barrier.wait();
+      f()
+    } else {
+      f()
+    }
+  }
+
+  pub fn terminate_loop_with_result(&self, result: Vec<RootResult>) {
+    *self.final_result.lock() = Some(result);
```

Comment:
It looks like rather than storing the `final_result` as a field of the struct, you could put it inside the `FinalResult` variant of the enum?

Reply:
Yeah that was the original version of the code; I moved it into the body of `ConsoleUI` to try to fix a bug and it didn't end up helping, so I'll move it back.